internal/extensionserver/post_cluster_modify.go

```diff
@@ -0,0 +1,385 @@
+	// Initialize cluster metadata structure if not present.
+	if cluster.Metadata == nil {
+		cluster.Metadata = &corev3.Metadata{}
+	}
+	if cluster.Metadata.FilterMetadata == nil {
+		cluster.Metadata.FilterMetadata = make(map[string]*structpb.Struct)
+	}
+
+	// Get or create the internal metadata namespace for AI Gateway.
+	m, ok := cluster.Metadata.FilterMetadata[internalapi.InternalEndpointMetadataNamespace]
+	if !ok {
+		m = &structpb.Struct{}
+		cluster.Metadata.FilterMetadata[internalapi.InternalEndpointMetadataNamespace] = m
+	}
+	if m.Fields == nil {
+		m.Fields = make(map[string]*structpb.Value)
+	}
+
+	// Store InferencePool reference as metadata for later retrieval.
+	// The reference includes all information needed to build EPP clusters and filters.
+	m.Fields[internalapi.InternalMetadataInferencePoolKey] = structpb.NewStringValue(
+		internalapi.ClusterRefInferencePool(
+			inferencePool.Namespace,
+			inferencePool.Name,
+			string(inferencePool.Spec.EndpointPickerConfig.ExtensionRef.Name),
+			portForInferencePool(inferencePool),
+		),
+	)
+}
+
+// getInferencePoolByMetadata returns the InferencePool from the cluster metadata.
+func getInferencePoolByMetadata(cluster *clusterv3.Cluster) *gwaiev1a2.InferencePool {
+	var metadata string
+	if cluster.Metadata != nil && cluster.Metadata.FilterMetadata != nil {
+		m, ok := cluster.Metadata.FilterMetadata[internalapi.InternalEndpointMetadataNamespace]
+		if ok && m.Fields != nil {
+			v, ok := m.Fields[internalapi.InternalMetadataInferencePoolKey]
+			if ok {
+				metadata = v.GetStringValue()
+			}
+		}
+	}
+
+	result := strings.Split(metadata, "/")
+	if len(result) != 4 {
+		return nil
+	}
+	ns := result[0]
+	name := result[1]
+	serviceName := result[2]
+	port, err := strconv.ParseInt(result[3], 10, 32)
+	if err != nil {
+		return nil
+	}
+	return &gwaiev1a2.InferencePool{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+		},
+		Spec: gwaiev1a2.InferencePoolSpec{
+			EndpointPickerConfig: gwaiev1a2.EndpointPickerConfig{
+				ExtensionRef: &gwaiev1a2.Extension{
+					ExtensionReference: gwaiev1a2.ExtensionReference{
+						Name:       gwaiev1a2.ObjectName(serviceName),
+						PortNumber: ptr.To(gwaiev1a2.PortNumber(port)),
+					},
+				},
+			},
+		},
+	}
+}
+
+// buildClustersForInferencePool builds and returns a "STRICT_DNS" cluster for each InferencePool cluster.
+func buildClustersForInferencePool(clusters []*clusterv3.Cluster) []*clusterv3.Cluster {
+	result := make([]*clusterv3.Cluster, 0, len(clusters))
+	for _, cluster := range clusters {
+		if pool := getInferencePoolByMetadata(cluster); pool != nil {
+			result = append(result, buildExtProcClusterForInferencePool(pool))
+		}
+	}
+
+	return result
+}
+
+// buildExtProcClusterForInferencePool builds and returns a "STRICT_DNS" cluster from the given pool.
+func buildExtProcClusterForInferencePool(pool *gwaiev1a2.InferencePool) *clusterv3.Cluster {
```

Comment:
let's make it more explicit about what this does since we have two kinds of clusters dealing with in this package regarding InferencePool. (one is for route and one is for endpoint picker)  

```suggestion
func buildExtProcClusterForInferencePoolEndpointPicker(pool *gwaiev1a2.InferencePool) *clusterv3.Cluster {
```